poc/vdaf_poc/flp_bbcggi19.py

```diff
@@ -275,11 +273,33 @@ def __init__(
             wire[0] = s  # set the wire seed
             self.wires.append(wire)
 
+        # Recover gadget_poly sent in the Lagrange basis.
+        l = len(gadget_poly)
+        n = next_power_of_2(l)
+        assert l <= n
+        lag = Lagrange(field)
+        nodes = field.nth_root_powers(n)
+        for _ in range(l, n):
+            # Extending the dimension (one by one) to the closest power of two.
+            # This keeps the degree of polynomial unchanged.
+            next_value = lag.extend_dimension_one(nodes, gadget_poly)
+            gadget_poly += [next_value]
+        # Calculate 'size' evaluations of gadget_poly at the roots of unity.
+        size = next_power_of_2(g.DEGREE*p)
```

Comment:
I think this is calculating more points than needed in edge cases. Note that the wire polynomials are all polynomials of degree at most (p-1), so the gadget polynomial has a degree of at most g.DEGREE * (p - 1), and therefore the following number of coordinates.
```suggestion
        size = next_power_of_2(g.DEGREE * (p - 1) + 1)
```

Reply:
That still works with the following change. Historically, there were a few places where we had overly conservative formulas for polynomial degrees like this, so I think we've just missed this one.

```patch
diff --git a/src/flp.rs b/src/flp.rs
index 7c85614..1153f76 100644
--- a/src/flp.rs
+++ b/src/flp.rs
@@ -717,7 +717,7 @@ impl<F: NttFriendlyFieldElement> QueryShimGadget<F> {
         let gadget_degree = inner.degree();
         let gadget_arity = inner.arity();
         let m = (1 + inner.calls()).next_power_of_two();
-        let p = m * gadget_degree;
+        let p = (m - 1) * gadget_degree + 1;
 
         // Each call to this gadget records the values at which intermediate proof polynomials were
         // interpolated. The first point was a random value chosen by the prover and transmitted in
```

In this Python code, `p = next_power_of_2(1 + g_calls)` is the number of coordinates that we use to define the wire polynomials, so the wire polynomials are of degree p-1 or less, and composition means the gadget polynomial will have degree `g.DEGREE * (p - 1)`. Thus, we need at least `g.DEGREE * (p - 1) + 1` coordinates for the gadget polynomial. Note that this also matches the calculation of `gadget_poly_len` in `prove()`. Here we round up to the next power of two, and later rely on that when getting roots of unity, but I'm not sure if that rounding up is strictly necessary.

Reply:
I did some further testing and experimentation of this in draft PR #589. One thing I discuss there is that the calculation of `step` will compensate for the extra basis extension in cases where `p` is larger than necessary. So, that explains why verification still works either way.

Reply:
The calculation is correct. As also noted in the [the paper](https://eprint.iacr.org/2019/188.pdf) the degree is at most `M*deg(G)`, where `M=p-1`. So the length of the coordinates vector is `g.degree*(p-1) + 1`.

<img width="987" height="357" alt="image" src="https://github.com/user-attachments/assets/0bb1dbc3-0e73-492f-bad2-f2b5e3e757bb" />